Virtual Remote Datasets

frcroth opened the pull request:
### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz (No frontend)

This PR adds routes in the datastore to access datasets by ID as opposed to by organizationId and directory. Datasources are cached in the datastore, the configuration is generated by WK from the DB. WK can also invalidate the DS cache when the dataset was changed. This PR only supports remote datasets, and especially no WKW datasets.

This adds new routes with the "wkDataset" path, since the dataset routes directly follow with the organizationId, making routes that use the datasetId there ambiguous. ~~Zarr routes are also duplicated to use dataset ids~~. In a followup, once wkw datasets are also supported, the old routes should be removed and the new ones can be renamed to have nicer sounding routes.


### Steps to test:
- In the two changed frontend files (rest_api.ts and wkstore_adapter.ts) change the datasetId to a remote dataset you have in your WK.
- Open a dataset (preferably the same dataset so that metadata matches) and see data, histogram.
- Explore a remote dataset as a zarr dataset (e.g. exploring http://localhost:9000/data/wkZarr/683d9643f600008d0e26a0a3)

### TODOs:
- [x] Cache invalidation
- [x] Remove todos
- [x] Revert frontend changes

### Issues:
- Contributes to #8384 

------
(Please delete unneeded items, merge only when none are left open)
- [ ] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Removed dev-only changes like prints and application.conf edits
- [ ] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [ ] Needs datastore update after deployment

fm3 commented on app/controllers/WKRemoteDataStoreController.scala:
```diff
@@ -259,6 +259,17 @@ class WKRemoteDataStoreController @Inject()(
       }
     }
 
+  def getDataset(name: String, key: String, datasetId: ObjectId): Action[AnyContent] =
```
```suggestion
  def getDataSource(name: String, key: String, datasetId: ObjectId): Action[AnyContent] =
```

(as it does not return a Dataset object, which is what is stored in postgres, I’d rather follow the DataSource terminology in this context)

fm3 commented on app/models/dataset/Dataset.scala:
```diff
@@ -1015,6 +1056,39 @@ class DatasetLastUsedTimesDAO @Inject()(sqlClient: SqlClient)(implicit ec: Execu
+  def parseRow(row: DatasetLayerAttachmentsRow): Fox[LayerAttachment] =
+    for {
+      dataFormat <- LayerAttachmentDataFormat.fromString(row.dataformat).toFox ?~> "Could not parse data format"
+      uri = new URI(row.path)
```
```suggestion
      uri <- tryo(new URI(row.path))
```

fm3 commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/AccessTokenService.scala:
```diff
@@ -20,7 +20,7 @@ object AccessMode extends ExtendedEnumeration {
 object AccessResourceType extends ExtendedEnumeration {
   type AccessResourceType = Value
-  val datasource, tracing, annotation, webknossos, jobExport = Value
+  val datasource, dataset, tracing, annotation, webknossos, jobExport = Value
```
hmm again naming things is hard :grimacing: I don’t really have a clean solution either. I hope we can just get rid of the traditional/datasource codepaths soon.

fm3 commented:
I talked to norman again and yes, we want to eventually only use the byId routes in the datastore. So the next steps (possibly in a follow-up PR) would be to duplicate all relevant datastore routes, and implement this for all dataset types, as well as switch the frontend to use only the new routes.

frcroth commented on app/models/dataset/DatasetService.scala:
```diff
@@ -274,6 +290,208 @@ class DatasetService @Inject()(organizationDAO: OrganizationDAO,
+                        df
+                      )
+                    case DataFormat.zarr3 =>
+                      Zarr3DataLayer(
+                        name,
+                        category,
+                        boundingBox,
+                        elementClass,
+                        mags,
+                        defaultViewConfiguration,
+                        adminViewConfiguration,
+                        coordinateTransformations,
+                        numChannels,
+                        additionalAxes,
+                        attachmentsOpt,
+                      )
```
No, trailing commas have been implemented in Scala 2.12.2 https://docs.scala-lang.org/sips/trailing-commas.html

frcroth commented:
@fm3 I changed it to use an AlfuCache and duplicated the zarr routes as discussed. The other routes are all about files so not relevant to this change (remote datasets). 

I also removed the frontend changes. If you want to test them again, see this commit here: https://github.com/scalableminds/webknossos/pull/8657/commits/f62a7d79c627b527389873680c361cc6eb69031e#diff-1833767de35826d9954d2ef7d077f41f8b69861833801e63b9dd77e47d4c6d27

fm3 reviewed (commented):
Nice, I think with the AlfuCache this makes a lot of sense :) Thanks for implementing the smaller comments as well.

I’m a little unsure about the new code duplication, see comment above.

fm3 commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/WKZarrStreamingController.scala:
```diff
@@ -0,0 +1,529 @@
+    remoteWebknossosClient: DSRemoteWebknossosClient,
+    remoteTracingstoreClient: DSRemoteTracingstoreClient,
+)(implicit ec: ExecutionContext)
+    extends Controller
+    with Zarr3OutputHelper {
+
+  override def defaultErrorCode: Int = NOT_FOUND
```
Or maybe we could go the other way round, and have the by-orgaId-and-datasetName routes do a lookup of the id, and then call the methods of the new controllers? If both lookups are cached on the datastore side, there should not be a significant performance hit for that, right? What do you think @frcroth ?
